test/e2e/clusterctl_upgrade.go

```diff
@@ -139,8 +148,27 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			clusterctlBinaryURLTemplate = input.InitWithBinary
 		}
 		clusterctlBinaryURLReplacer := strings.NewReplacer("{OS}", runtime.GOOS, "{ARCH}", runtime.GOARCH)
-		clusterctlBinaryURL = clusterctlBinaryURLReplacer.Replace(clusterctlBinaryURLTemplate)
-		Expect(input.E2EConfig.Variables).To(HaveKey(initWithKubernetesVersion))
+		initClusterctlBinaryURL = clusterctlBinaryURLReplacer.Replace(clusterctlBinaryURLTemplate)
+
+		// NOTE: by default we are considering all the providers, no matter of the contract.
+		// However, given that we want to test both v1alpha3 --> v1beta1 and v1alpha4 --> v1beta1, the INIT_WITH_PROVIDERS_CONTRACT
+		// variable can be used to select versions with a specific contract.
+		initContract = "*"
+		if input.E2EConfig.HasVariable(initWithProvidersContract) {
+			initContract = input.E2EConfig.GetVariable(initWithProvidersContract)
+		}
+		if input.InitWithProvidersContract != "" {
+			initContract = input.InitWithProvidersContract
+		}
+
+		if input.InitWithKubernetesVersion == "" {
+			Expect(input.E2EConfig.Variables).To(HaveKey(initWithKubernetesVersion), "Invalid argument. %s variable must be defined when calling %s spec", initWithKubernetesVersion, specName)
+			Expect(input.E2EConfig.Variables[initWithKubernetesVersion]).ToNot(BeEmpty(), "Invalid argument. %s variable can't be empty when calling %s spec", initWithKubernetesVersion, specName)
+			initKubernetesVersion = input.E2EConfig.GetVariable(initWithKubernetesVersion)
+		} else {
+			initKubernetesVersion = input.InitWithKubernetesVersion
+		}
```

Comment:
nit (to avoid the else)
```suggestion
		initKubernetesVersion = input.InitWithKubernetesVersion
		if input.InitWithKubernetesVersion == "" {
			Expect(input.E2EConfig.Variables).To(HaveKey(initWithKubernetesVersion), "Invalid argument. %s variable must be defined when calling %s spec", initWithKubernetesVersion, specName)
			Expect(input.E2EConfig.Variables[initWithKubernetesVersion]).ToNot(BeEmpty(), "Invalid argument. %s variable can't be empty when calling %s spec", initWithKubernetesVersion, specName)
			initKubernetesVersion = input.E2EConfig.GetVariable(initWithKubernetesVersion)
		} 
```

Reply:
Good idea!